InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Validations.Modify.cs

```diff
@@ -0,0 +1,380 @@
+
+            this.dateTimeBrokerMock.VerifyNoOtherCalls();
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+            this.storageBrokerMock.VerifyNoOtherCalls();
+        }
+
+        [Fact]
+        public async Task ShouldThrowValidationExceptionOnModifyIfInternDoesntExistAndLogItAsync()
+        {
+            DateTimeOffset randomDateTimeOffset = GetRandomDateTimeOffset();
+            Intern nonExistentIntern = CreateRandomIntern();
+            nonExistentIntern.UpdatedDate = randomDateTimeOffset;
+            Intern noIntern = null;
+            var notFoundInternException = new NotFoundInternException(nonExistentIntern.Id);
+
+            var expectedInternValidationException =
+                new InternValidationException(notFoundInternException);
+
+            this.storageBrokerMock.Setup(broker =>
+                broker.SelectInternByIdAsync(nonExistentIntern.Id))
+                    .ReturnsAsync(noIntern);
+
+            this.dateTimeBrokerMock.Setup(broker =>
+                broker.GetCurrentDateTimeOffset())
+                    .Returns(randomDateTimeOffset);
+
+            // when
+            ValueTask<Intern> modifyInternTask =
+                this.internService.ModifyInternAsync(nonExistentIntern);
+
+            InternValidationException actualInternValidationException =
+                await Assert.ThrowsAsync<InternValidationException>(
+                    modifyInternTask.AsTask);
+
+            // then
+            actualInternValidationException.Should().BeEquivalentTo(
+                expectedInternValidationException);
+
+            this.dateTimeBrokerMock.Verify(broker =>
+                broker.GetCurrentDateTimeOffset(),
+                    Times.Once);
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogError(It.Is(SameExceptionsAs(
+                    expectedInternValidationException))),
+                        Times.Once);
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.SelectInternByIdAsync(nonExistentIntern.Id),
+                    Times.Once);
+
+            this.dateTimeBrokerMock.VerifyNoOtherCalls();
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+            this.storageBrokerMock.VerifyNoOtherCalls();
+        }
+
+        [Fact]
+        public async Task 
+            ShouldThrowValidationExceptionOnModifyIfStorageInternAuditInfoNotSameAsInputInternAuditInfoAndLogItAsync()
+        {
+            // given
+            int randomNumber = GetRandomNumber();
+            DateTimeOffset randomDate = GetRandomDateTime();
+            Guid differentId = Guid.NewGuid();
+            Guid invalidCreatedBy = differentId;
+            Intern randomIntern = CreateRandomIntern(dates: randomDate);
+            Intern invalidIntern = randomIntern;
+            Intern storageIntern = randomIntern.DeepClone();
+            invalidIntern.CreatedDate = storageIntern.CreatedDate.AddDays(randomNumber);
```

Comment:
Does this pass use-case when we check CreatedDate is earlier than UpdatedDate?

Reply:
You should setup your values so that existing logic doesn't error out.

Reply:
@Pwongchaiya assume that your random intern was created with random date `1/1/2021`
If `randomNumber` is `9` then your new `CreatedDate` is `1/10/2021`
But that means that your `UpdatedDate` is still `1/1/2021`
And THAT should failure validation